### DISCUSSION TITLE
Native integration of job scheduling

### DISCUSSION THREAD

**Discussion by maietta** [caprover/caprover#1530] [Ideas]
I maintain a docker-oriented job scheduler called Chadburn found here: https://github.com/PremoWeb/chadburn.

I've been mulling over the idea of the best way to use Chadburn with CapRover and currently have a One-Click-App which works great, but with some caveats. Mainly, renaming apps causes the abandonment of jobs due to the nature of the way container labels to work.

This morning, a user-contributed feature was added for Prometheus metric endpoint monitoring via an optional CLI flag. The use of the One-Click-App for deploying Chadburn is limited in that I don't have a clean way to optionally enable or disable this metric endpoint. (If there's a way, I'd love to know)

I envision that a future CapRover would offer built-in scheduling of tasks on containers, much like a CRON job. Being able to visit an app and set up scheduled tasks would be awesome. Being able to also enable Prometheus monitoring endpoints with the click of a button would also be great.

I don't know if it makes sense to have Chadburn integrated with CapRover, but I want to hear what others think.

**Comment by maietta** [comment]
Chadburn listens to changes to container state and reads labels for tasks, so no need to directly notify Chadburn of changed states.

Some notable tasks that would have to be performed as part of the integration include updating the container labels when an app name is updated when an app instance count is modified (0 vs anything more than 0).

For apps with multiple instances, there's a job overlap flag that can be set to true or false.

Jobs can be executed in 4 different ways:

- On a container already running
- On a new container
- On the host
- Service executed on new container

While Chadburn does listen for container labels, it can also accept an INI formatted config file. In fact, if both are available, the configs are merged. Meaning, you can predefine an INI config as part of the host config and then optionally offer container labels per app.

Chadburn also offers in-built notification options including Email, Slack, Microsoft Teams, and Gotify webhook methods, or even save notifications to a directory, useful for keeping tabs on scheduled tasks.

Seems like Chadburn could be a nice feature addition if integrated with the API and UI.

**Comment by waschinski** [comment]
With Chadburn now being archived I'd like to support this old request.
